### PR TITLE
Feature 339 tcdiag skewt

### PR DESCRIPTION
**Pull request by bikegeek**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
 
   on host 'seneca', clone the feature_339_tcdiag_skewt branch: 

-        git clone https://github.com/dtcenter/METplotpy
-        git checkout feature_339_tcdiat_skewt
-        set the METPLOTPY_BASE env: 
-                 export METPLOTPY_BASE=/d1/personal/mwin/feature_339_tcdidag_skewt

   
   change to 'bash' shell, from command line enter:  

-         bash

      

   set the PYTHONPATH: 

-          export PYTHONPATH=$METPLOTPY_BASE/METplotpy:$METPLOTPY_BASE/METplotpy/metplotpy:$METPLOTPY_BASE/METplotpy/plots 

Re-create the Python environment:

-    cd $METPLOTPY_BASE/METplotpy/metplotpy/plots/skew_t
-    conda env create -f tcdiag_env.yaml

   using the data files in the following directory on 'seneca' (this will be used in the config file):

-     /d1/projects/TCDiag/data_output/DIAGNOSTICS/TCDIAG/2023

   modified the skew_t.yaml config file to define to the log directory, log filename, input file directory, output file directory

cd to $METPLOTPY_BASE/METplotpy/metplotpy/plots/skew_t

Modify the skew_t.yaml configuration file:

-     set the input directory to /d1/projects/TCDiag/data_output/DIAGNOSTICS/TCDIAG/2023/sh162023
-     set the output directory to wherever you wish
-     set the log_directory to wherever you wish
-     change the log_filename to whatever you wish, default is tc_diags_skewt.log
current settings are for: 
-            only 0 and 6 hour soundings
-            display all the adiabats, mixing lines, and wind barbs (display_xyz: True, where xyz=dry_adiabats, moist_adiabats, etc.)
-            wind barbs thinned out by setting decimate to 2 (every two wind barbs will be displayed)
-            pressure levels displayed adjacent to the y2 axis (level_labels_along_y2-axis: True)

       
Generate the plot. From the command line:
   python skew_t.py skew_t.yaml

Verified that the plots generated have the expected lines, wind barbs, pressure levels along the y2-axis.  

Plots should look like this (title contains the information extracted from the input data file name):
Corresponds to the data from 2023/sh052023/ssh052023_avno_doper_2023010100_diag.dat 

ssh052023_avno_doper_2023010100_diag_0_hr

    
       
    

 

  Modified the config file to plot the pressure levels adjacent to the temperature line, turned off wind barb decimation, changed plot size, change line colors for temp, etc.  Verified the expected changes appear on the plot.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

    Perform the same steps above (changing directory path to $METPLOTPY_BASE) 
    Verify that wind barb decimating/resampling works (1=no resampling) and the pressure level labels can be displayed either along the y2-axis or adjacent to the temperature curve
 
     *Note:  There currently isn't any error-checking for invalid data values (i.e. 9999).  

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[ No]**  Documentation in ReadTheDocs will need to be created.

- [x] Do these changes include sufficient testing updates? **[No]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Please complete this pull request review by **Monday June 12**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by jvigh** [comment]
Here's one of my more wacky tests, in which I changed many config options:
image

**Comment by jvigh** [comment]
There was at least one typo in the pull request testing instructions, which threw me off, so I created a shell script to work through each step. In the process, I came up with something that might be able to be applied more generally to pull request testing of any METplotpy capability.  Note that it departs a bit from Minna's suggested directly structure in that it doesn't include the extra METplotpy in the path. This is so that multiple branches can be tested and the results kept for future reference (if desired). 

TEST_METplotpy_feature.bash:

```bash

#!/bin/bash
# script TEST_METplotpy_feature.bash"

# This script conducts pull request testing for METplotpy development issues.
# While it has been generalized to ease testing for any METplotpy PR, it has only been
# tested on PR #352 so far.

# Author: Jonathan Vigh, 08 June 2023

echo ""
echo "=========================================================="
echo " Welcome to METplotpy's generalized pull request tester!"
echo "=========================================================="
echo ""


# Modify the following parameters as needed
PROJECT_BASE="/d1/projects/TCDiag"
MET_COMPONENT="METplotpy"
GIT_BRANCH="feature_339_tcdiag_skewt"
TEST_DIR="${MET_COMPONENT}_${GIT_BRANCH}"
TEST_DATA="${PROJECT_BASE}/data_output/DIAGNOSTICS/TCDIAG/2023"
CREATE_CONDA_ENV="False"
PLOT_TYPE="skew_t"


# Set up the repo and checkout the desired feature branch to be tested
if [[ ! -d ${TEST_DIR} ]]
then
  cd ${PROJECT_BASE}
  git clone git@github.com:dtcenter/METplotpy.git ${TEST_DIR}
  cd ${PROJECT_BASE}/${TEST_DIR}
  git checkout ${GIT_BRANCH}
fi

export METPLOTPY_BASE=${PROJECT_BASE}/${TEST_DIR}
export PYTHONPATH=${METPLOTPY_BASE}/:${METPLOTPY_BASE}/metplotpy:${METPLOTPY_BASE}/plots


# Recreate the python environment
if [[ ${CREATE_CONDA_ENV} == "True" ]]
then
  cd ${METPLOTPY_BASE}/metplotpy/plots/${PLOT_TYPE}
  conda env create -f tcdiag_env.yaml
fi


# Activate the python environment
source /var/autofs/mnt/linux-amd64/debian/buster/local/anaconda3-20230129/etc/profile.d/conda.sh
conda activate tcdiag_env
echo ""
echo "Checking to see which conda environment is active (the one with the *):"
conda info --env


# Check test data
echo ""
echo "Checking the contents of the test data directory:"
ls ${TEST_DATA}


# Now display the README to provide the user with further instructions
echo ""
echo "Now changing to the configuration directory:"
cd ${METPLOTPY_BASE}/metplotpy/plots/${PLOT_TYPE}
pwd
echo ""
echo "This directory has the following contents:"
ls
echo ""
echo "This directory contains the configuration files for the capability to be tested."
echo "Further instructions are provided in the README file."


# Set up test config file by copying the original (only if the test file has not already been created previously, to prevent overwriting user's existing test config)
if [[ ! -s ${PLOT_TYPE}_TEST.yaml ]]
then
  echo "Copying ${PLOT_TYPE}.yaml to ${PLOT_TYPE}_TEST.yaml"
  cp ${PLOT_TYPE}.yaml ${PLOT_TYPE}_TEST.yaml
  sleep 4
fi


# Now run the plotter
echo ""
echo "Now running the ${PLOT_TYPE} plotter for ${PLOT_TYPE}_TEST.yaml . . ."
echo ""
echo "python ${PLOT_TYPE}.py ${PLOT_TYPE}_TEST.yaml"
python ${PLOT_TYPE}.py ${PLOT_TYPE}_TEST.yaml
echo ""
echo "To make changes to the test configuration for this plotter, please modify ${PLOT_TYPE}_TEST.yaml and reun this script."
echo ""
pwd
```

File: metplotpy/plots/skew_t/skew_t.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
+# ============================*
+# ** Copyright UCAR (c) 2020
```
Comment (jvigh) [inline metplotpy/plots/skew_t/skew_t.py]: Change 2020 to 2023?

File: metplotpy/plots/skew_t/skew_t.py [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+def retrieve_units(sounding_data: pd.DataFrame) -> dict:
+    '''
+       Retrieve the units from the ASCII file.
+
+       Args:
+          sounding_data:  The pandas dataframe containing the sounding data.
+
+       Returns:
+         units: a dictionary of the field as key and units as value (where the
+         parenthesis
```
Comment (jvigh) [inline metplotpy/plots/skew_t/skew_t.py:121]: Just an odd line end here.

File: metplotpy/plots/skew_t/skew_t.py [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+    for cur_lev in plevs:
+        # Create the string describing the pressure levels
+        temp_prefix = 'T_'
+        temp_field = temp_prefix + cur_lev
+
+        # Get the location of the temperature from the temp_field row.
+        temp_idx: pd.Index = sounding_data.index[sounding_data.FIELD == temp_field]
+        hours: pd.Series = sounding_data[hour_of_interest]
+        temp_hour: pd.Series = hours.iloc[temp_idx].values[0]
+        # Temperatures need to be divie by 10 since units are 10C
```
Comment (jvigh) [inline metplotpy/plots/skew_t/skew_t.py]: divie -> divided

File: metplotpy/plots/skew_t/skew_t.py [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+                                                            cur_rh * units.percent)
+            dewpoint = calc_dewpoint.magnitude
+        dewpts.append(dewpoint)
+
+    return dewpts
+
+
+def retrieve_height(sounding_data: pd.DataFrame, hour_of_interest: str,
+                    all_pressure_levels: list) -> list:
+    '''Retrieve all the hsights (in decameters) in the soundings dataframe based on
```
Comment (jvigh) [inline metplotpy/plots/skew_t/skew_t.py]: hsights -> heights

File: metplotpy/plots/skew_t/skew_t.py [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+        hgt_idx: pd.Index = sounding_data.index[sounding_data.FIELD == hgt_field]
+        hours: pd.Series = sounding_data[hour_of_interest]
+        # Convert from decameters to meters
+
+        #  Make special accomodation for no Z_SURF, which is just 0
+        if hgt_field == 'Z_SURF':
+            hgt_hour = 0
+        else:
+            try:
+
```
Comment (jvigh) [inline metplotpy/plots/skew_t/skew_t.py]: Should this blank line be here?

**Review by jvigh — APPROVED** [review]
@bikegeek  Here's a summarized review.

As per my previous comments, I have exhaustively tested this capability and have not identified any usability issues. I have tested each option, either in isolation or together with other options. They all seem to work fine. The output looks good. I understand that the documentation will be added later (or is there something I should review?). 

My substantive comment is to offer a suggestion on the handling for the TCDiag missing values (9999). This would be to check ALL values for a given time. If they are all missing, then don't error out, but print a message on the plot that "All data are missing for this time." If some data are good but others are bad for a given time slice, then print the data that are good and ignore the bad data (by setting to NaN or whatever is prudent in python). 

I have run the code one last time using the latest code. It still works. Therefore, I approve this pull request.

This is the first graphical capability and a nice addition to what I'm sure will become a useful suite of capabilities for future users. Good work!

**Comment by bikegeek** [comment]
Added checking for data with no data (all missing data) and data with partial missing data.  The code now logs warning messages and moves to the next file (if current file is completely bereft of data) or moves to the next hour of data if a particular hour and variable (temperature or relative humidity) are completely full of missing data.  A test has been written using data files that have some missing data or are completely full of missing data.

**Review by jvigh — APPROVED** [review]
@bikegeek  Thanks for adding the error checking for bad data values. I have reviewed the code changes. I have also tested the new code on the test data in the METplotpy/test/skew_t/data (NOTE: I used my script, rather than the python test script, as I don't know how to run python tests yet.). 

For sh162023, 44 files were generated and none of them were from data files with missing data. For sh052023, 4 files were generated. 

I am not sure if skipping over the files with missing data will be suitable, as real-time and display systems may still expect a file to display. That is why I had suggested putting a message on the plot saying "No data are available for plotting." or something to that effect. 

I have written to Kate Musgrave to ask her opinion. In the meantime, if this needs to go for the release tomorrow, I think it's fine to proceed. The behavior of what to do with missing data can be improved in a later update.

Thanks!

**Review by hankenstein2 — APPROVED** [review]
I went through @jvigh testing instructions and was able to test several plot variations as well as the minor code text changes.  Looks great Minna

**Pull request merged**